### PR TITLE
bump(x11/audacity): 3.6.4

### LINKED ISSUE #5422: Need Help: Sound recording issue with pyaudio
**Problem description**

Created a pyaudio script t.py to record sound and export it as wav file.
Problem:In termux the file shows that it started recording but strucks at it and does not stop and requires stopping script with ctrl+z while ctrl+c did not workit does not write any recording to wav. Means no recording.  The script is made to record 5 seconds of audio

Script content:
```
import pyaudio
import wave

FORMAT = pyaudio.paInt16
CHANNELS = 2
RATE = 44100
CHUNK = 1024
RECORD_SECONDS = 5
WAVE_OUTPUT_FILENAME = "file.wav"

audio = pyaudio.PyAudio()

# start Recording
stream = audio.open(format=FORMAT, channels=CHANNELS,
                rate=RATE, input=True,
                frames_per_buffer=CHUNK)
print ("recording...")

frames = []

for i in range(0, int(RATE / CHUNK * RECORD_SECONDS)):
    data = stream.read(CHUNK)
    frames.append(data)
print ("finished recording")


# stop Recording
stream.stop_stream()
stream.close()
audio.terminate()

waveFile = wave.open(WAVE_OUTPUT_FILENAME, 'wb')
waveFile.setnchannels(CHANNELS)
waveFile.setsampwidth(audio.get_sample_size(FORMAT))
waveFile.setframerate(RATE)
waveFile.writeframes(b''.join(frames))
waveFile.close()
```
**Steps to reproduce**

**Expected behavior**

Recording of 5 seconds audio in wav file
**Additional information**

Tried solution:
Installed ubuntu on termux with vnc server x11 working and on it tried to run the script,  and following error is obtained: 
Error log from ubuntu installed on termux is  given below
```
ALSA lib confmisc.c:767:(parse_card) cannot find card '0'
ALSA lib conf.c:4693:(_snd_config_evaluate) function snd_func_card_driver returned error: No such file or directory
ALSA lib confmisc.c:392:(snd_func_concat) error evaluating strings
ALSA lib conf.c:4693:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1246:(snd_func_refer) error evaluating name
ALSA lib conf.c:4693:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5181:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM sysdefault
ALSA lib confmisc.c:767:(parse_card) cannot find card '0'
ALSA lib conf.c:4693:(_snd_config_evaluate) function snd_func_card_driver returned error: No such file or directory
ALSA lib confmisc.c:392:(snd_func_concat) error evaluating strings
ALSA lib conf.c:4693:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1246:(snd_func_refer) error evaluating name
ALSA lib conf.c:4693:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5181:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM sysdefault
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.front
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.rear
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.center_lfe
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.side
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.surround21
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.surround21
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.surround40
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.surround41
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.surround50
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.surround51
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.surround71
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.iec958
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.iec958
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.iec958
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.hdmi
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.hdmi
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.modem
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.modem
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.phoneline
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.phoneline
ALSA lib pcm_hw.c:1829:(_snd_pcm_hw_open) Invalid value for card
ALSA lib pcm_hw.c:1829:(_snd_pcm_hw_open) Invalid value for card
ALSA lib pcm_hw.c:1829:(_snd_pcm_hw_open) Invalid value for card
ALSA lib pcm_hw.c:1829:(_snd_pcm_hw_open) Invalid value for card
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
ALSA lib pcm_oss.c:377:(_snd_pcm_oss_open) Unknown field port
ALSA lib pcm_oss.c:377:(_snd_pcm_oss_open) Unknown field port
ALSA lib pulse.c:242:(pulse_connect) PulseAudio: Unable to connect: Access denied

ALSA lib pulse.c:242:(pulse_connect) PulseAudio: Unable to connect: Access denied

ALSA lib pcm_hw.c:1829:(_snd_pcm_hw_open) Invalid value for card
ALSA lib pcm_hw.c:1829:(_snd_pcm_hw_open) Invalid value for card
ALSA lib pcm_hw.c:1829:(_snd_pcm_hw_open) Invalid value for card
ALSA lib pcm_hw.c:1829:(_snd_pcm_hw_open) Invalid value for card
ALSA lib pcm_usb_stream.c:486:(_snd_pcm_usb_stream_open) Invalid type for card
ALSA lib pcm_usb_stream.c:486:(_snd_pcm_usb_stream_open) Invalid type for card
ALSA lib pulse.c:242:(pulse_connect) PulseAudio: Unable to connect: Access denied

ALSA lib pulse.c:242:(pulse_connect) PulseAudio: Unable to connect: Access denied

ALSA lib confmisc.c:767:(parse_card) cannot find card '0'
ALSA lib conf.c:4693:(_snd_config_evaluate) function snd_func_card_driver returned error: No such file or directory
ALSA lib confmisc.c:392:(snd_func_concat) error evaluating strings
ALSA lib conf.c:4693:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1246:(snd_func_refer) error evaluating name
ALSA lib conf.c:4693:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5181:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM dmix
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
Traceback (most recent call last):
  File "t.py", line 14, in <module>
    stream = audio.open(format=FORMAT, channels=CHANNELS,
  File "/usr/lib/python3/dist-packages/pyaudio.py", line 750, in open
    stream = Stream(self, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/pyaudio.py", line 441, in __init__
    self._stream = pa.open(**arguments)
OSError: [Errno -9996] Invalid input device (no default output device)
```

Then tried another solution
 dbus-launch --exit-with-session pulseaudio --start --daemon as non root 
also tried
dbus-launch --exit-with-session pulseaudio --start --daemon --system --disallow-exit as root user
python t.py
This time script ran and stopped after 5 seconds and file.wav is also created but it was mute no audio was actually recorded.

And output is 
```
ALSA lib confmisc.c:767:(parse_card) cannot find card '0'
ALSA lib conf.c:4693:(_snd_config_evaluate) function snd_func_card_driver returned error: No such file or directory
ALSA lib confmisc.c:392:(snd_func_concat) error evaluating strings
ALSA lib conf.c:4693:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1246:(snd_func_refer) error evaluating name
ALSA lib conf.c:4693:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5181:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM sysdefault
ALSA lib confmisc.c:767:(parse_card) cannot find card '0'
ALSA lib conf.c:4693:(_snd_config_evaluate) function snd_func_card_driver returned error: No such file or directory
ALSA lib confmisc.c:392:(snd_func_concat) error evaluating strings
ALSA lib conf.c:4693:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1246:(snd_func_refer) error evaluating name
ALSA lib conf.c:4693:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5181:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM sysdefault
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.front
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.rear
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.center_lfe
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.side
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.surround21
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.surround21
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.surround40
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.surround41
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.surround50
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.surround51
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.surround71
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.iec958
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.iec958
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.iec958
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.hdmi
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.hdmi
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.modem
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.modem
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.phoneline
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.phoneline
ALSA lib pcm_hw.c:1829:(_snd_pcm_hw_open) Invalid value for card
ALSA lib pcm_hw.c:1829:(_snd_pcm_hw_open) Invalid value for card
ALSA lib pcm_hw.c:1829:(_snd_pcm_hw_open) Invalid value for card
ALSA lib pcm_hw.c:1829:(_snd_pcm_hw_open) Invalid value for card
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
ALSA lib pcm_oss.c:377:(_snd_pcm_oss_open) Unknown field port
ALSA lib pcm_oss.c:377:(_snd_pcm_oss_open) Unknown field port
ALSA lib pcm_hw.c:1829:(_snd_pcm_hw_open) Invalid value for card
ALSA lib pcm_hw.c:1829:(_snd_pcm_hw_open) Invalid value for card
ALSA lib pcm_hw.c:1829:(_snd_pcm_hw_open) Invalid value for card
ALSA lib pcm_hw.c:1829:(_snd_pcm_hw_open) Invalid value for card
ALSA lib pcm_usb_stream.c:486:(_snd_pcm_usb_stream_open) Invalid type for card
ALSA lib pcm_usb_stream.c:486:(_snd_pcm_usb_stream_open) Invalid type for card
ALSA lib confmisc.c:767:(parse_card) cannot find card '0'
ALSA lib conf.c:4693:(_snd_config_evaluate) function snd_func_card_driver returned error: No such file or directory
ALSA lib confmisc.c:392:(snd_func_concat) error evaluating strings
ALSA lib conf.c:4693:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1246:(snd_func_refer) error evaluating name
ALSA lib conf.c:4693:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5181:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM dmix
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
recording...
finished recording
```
ls results shown file.wav but it was having no sound.

However on termux script keeps working infinitely and does not stop after 5 seconds and we need to force stop it with ctrl+z, while with ubuntu on termux script rans records and stops and creates output but is muted or no actual sound recorded
Termux is been provided with microphone access permission using termux-microphone-record api.

Other solutions tried:
sudo jackd -d alsa to run jack server but in ended with following error:
```jackdmp 1.9.12
Copyright 2001-2005 Paul Davis and others.
Copyright 2004-2016 Grame.
Copyright 2016-2017 Filipe Coelho.
jackdmp comes with ABSOLUTELY NO WARRANTY
This is free software, and you are welcome to redistribute it
under certain conditions; see the file COPYING for details
JACK semaphore error: semget creation (Function not implemented)
jack_shm_lock_registry fails...
No access to shm registry
Failed to open server```

### LINKED ISSUE #10412: Audacity does not load/detect pulseaudio or any audio devices
### Problem description

I got this issue where audacity does not detect pulseaudio 
audacity-err

### What steps will reproduce the bug?

Open audacity 

### What is the expected behavior?

Pulseaudio or opensles loaded as audio device

### System information

termux-info:

```
Termux Variables:
TERMUX_API_VERSION=0.50.1
TERMUX_APK_RELEASE=F_DROID
TERMUX_APP_PID=26202
TERMUX_IS_DEBUGGABLE_BUILD=0
TERMUX_MAIN_PACKAGE_FORMAT=debian
TERMUX_VERSION=0.118.0
Packages CPU architecture:
aarch64
Subscribed repositories:
# sources.list
deb https://grimler.se/termux-packages-24/ stable main
# x11-repo (sources.list.d/x11.list)
deb https://packages-cf.termux.org/apt/termux-x11 x11 main
# sources.list.d/pointless.list
deb https://its-pointless.github.io/files/24 termux extras
Updatable packages:
All packages up to date
Android version:
9
Kernel build information:
Linux localhost 4.4.153-perf+ #1 SMP PREEMPT Thu Feb 20 17:26:22 CST 2020 aarch64 Android
Device manufacturer:
asus
Device model:
ASUS_X00TD

```

### LINKED ISSUE #18790: [Bug]: Audacity can not be rebuilt or upgraded.
### Problem description

Hi. Audacity can not be rebuilt, probably because of NDK verbump.
[Build log](https://github.com/termux/termux-packages/files/13752302/audacity-build.log).

Version upgrading is not possible too because of major changes in Audacity and patches incompatibility.

### What steps will reproduce the bug?

```
./build-package -I -f audacity
```

### LINKED ISSUE #19679: [Question] How to use the host device microphone in Audacity
Screenshot_2024-04-06_03-11-39

Can Audacity installed on Termux use the host device's microphone?

Since `termux-api` supports `termux-microphone-record`, I think there must be some way.

I wonder if it is possible.

### LINKED ISSUE #20570: [Bug]: Python has no access to audio devices
### Problem description

Sox's `rec` and `play` working good, but from python modules can`t access to devices.
I'm attempt to use sounddevice and pyaudio. Devices list is empty. 
```
import sounddevice as sd

sd.get_portaudio_version()
# (1246976, 'PortAudio V19.7.0-devel, revision 147dd722548358763a8b649b3e4b41dfffbcfbb6')
sd.query_devices()
# <empty responce>
```

```
import pyaudio
p = pyaudio.PyAudio()
p.get_device_count()
# 0
```

```
import speech_recognition as sr
sr.Microphone.list_microphone_names()
# []
```

### What steps will reproduce the bug?

pkg update
pkg upgrade
pkg i termux-api `pkg i termux-api`
pkg i portaudio `(19.07.00-1)`
pkg i pulseaudio `(17.0-1)`
pkg i sox `(14.4.2-25)`
pkg i python `(3.11.9-2)`
pip -V
`pip 24.0 from /data/data/com.termux/files/usr/lib/python3.11/site-packages/pip (python 3.11`
pip install --upgrade setuptools wheel
`setuptools-70.0.0-py3-none-any.whl` `wheel-0.43.0-py3-none-any.whl`

pip install sounddevice pyaudio
`sounddevice-0.4.7-py3-none-any.whl` `PyAudio-0.2.14.tar.gz`
pip list
```
Package     Version
----------- -------
cffi        1.16.0
pip         24.0
PyAudio     0.2.14
pycparser   2.22
setuptools  70.0.0
sounddevice 0.4.7
wheel       0.43.0
```

`<restarting termux>`
pulsesound --start
pactl load-module module-sles-source
`17`
pactl list sources short
```
1       OpenSL_ES_sink.monitor  module-sles-sink.c      s16le 2ch 44100Hz       IDLE
2       OpenSL_ES_source        module-sles-source.c    s16le 2ch 44100Hz       IDLE
```

rec 'test.wav'
```
Input File     : 'default' (pulseaudio)
Channels       : 2
Sample Rate    : 48000
Precision      : 16-bit
Sample Encoding: 16-bit Signed Integer PCM

In:0.00% 00:00:03.41 [00:00:00.00] Out:160k  [======|====  ] Hd:0.0 Clip:7 
```
play 'test.wav'
```
test.wav:

 File Size: 999k      Bit Rate: 1.54M
  Encoding: Signed PCM    
  Channels: 2 @ 16-bit   
Samplerate: 48000Hz      
Replaygain: off         
  Duration: 00:00:05.21  

In:18.0% 00:00:00.94 [00:00:04.27] Out:45.1k [  -===|==-   ]        Clip:0 
```

`<voice recorded and played>`

python
```
>>> import sounddevice as sd
>>> sd.query_devices()

>>> sd.query_devices(-1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/sounddevice.py", line 569, in query_devices
    raise PortAudioError(f'Error querying device {device}')
sounddevice.PortAudioError: Error querying device -1
>>> sd.query_devices(2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/sounddevice.py", line 569, in query_devices
    raise PortAudioError(f'Error querying device {device}')
sounddevice.PortAudioError: Error querying device 2
>>> sd.query_devices(1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/sounddevice.py", line 569, in query_devices
    raise PortAudioError(f'Error querying device {device}')
sounddevice.PortAudioError: Error querying device 1

```

### What is the expected behavior?

_No response_

### System information

tried at two device (android 9 and 13), termux 0.118:
```shell
Termux Variables:
TERMUX_API_VERSION=0.50.0
TERMUX_APK_RELEASE=GITHUB
TERMUX_APP_PACKAGE_MANAGER=apt
TERMUX_APP_PID=3519
TERMUX_IS_DEBUGGABLE_BUILD=1
TERMUX_MAIN_PACKAGE_FORMAT=debian
TERMUX_VERSION=0.118.0
TERMUX__USER_ID=0
Packages CPU architecture:
aarch64
Subscribed repositories:
# sources.list
deb http://mirror.mephi.ru/termux/termux-main stable main
Updatable packages:
All packages up to date
termux-tools version:
1.42.4
Android version:
9
Kernel build information:
Linux localhost 4.4.156-OniiChanKernel-R3+ #2 SMP PREEMPT Thu Sep 5 02:57:49 UTC 2019 armv8l Android
Device manufacturer:
Xiaomi
Device model:
Redmi Note 7
LD Variables:
LD_LIBRARY_PATH=
LD_PRELOAD=/data/data/com.termux/files/usr/lib/libtermux-exec.so
Installed termux plugins:
com.termux.api versionCode:50
```

```
Termux Variables:
TERMUX_API_VERSION=0.50.0
TERMUX_APK_RELEASE=GITHUB
TERMUX_APP_PACKAGE_MANAGER=apt
TERMUX_APP_PID=13496
TERMUX_IS_DEBUGGABLE_BUILD=1
TERMUX_MAIN_PACKAGE_FORMAT=debian
TERMUX_VERSION=0.118.0
TERMUX__USER_ID=0
Packages CPU architecture:
aarch64
Subscribed repositories:
# sources.list
deb http://mirror.mephi.ru/termux/termux-main stable main
# x11-repo (sources.list.d/x11.list)
deb http://mirror.mephi.ru/termux/termux-x11 x11 main
# root-repo (sources.list.d/root.list)
deb http://mirror.mephi.ru/termux/termux-root root stable
Updatable packages:
All packages up to date
termux-tools version:
1.42.4
Android version:
13
Kernel build information:
Linux localhost 4.14.328-perf-@b28c10f6-dirty #1 SMP PREEMPT Mon Feb 26 14:50:07 CET 2024 aarch64 Android
Device manufacturer:
Xiaomi
Device model:
M2007J20CG
LD Variables:
LD_LIBRARY_PATH=
LD_PRELOAD=/data/data/com.termux/files/usr/lib/libtermux-exec.so
Installed termux plugins:
com.termux.api versionCode:50
```

### LINKED ISSUE #21545: [Bug]:Portaudio couldn't detect audio device
### Problem description

I tried to use pyaudio library to record my audio.But after initialising pulse audio server and giving required permissions to termux api it doesn't work.

### What steps will reproduce the bug?

Install portaudio version(19.07) from apt and install pyaudio library.Then executing this code
```
import pyaudio
import wave

chunk = 1024  # Record in chunks of 1024 samples
sample_format = pyaudio.paInt16  # 16 bits per sample
channels = 2
fs = 44100  # Record at 44100 samples per second
seconds = 3
filename = "output.wav"

p = pyaudio.PyAudio()  # Create an interface to PortAudio

print('Recording')

stream = p.open(format=sample_format,
                channels=channels,
                rate=fs,
                frames_per_buffer=chunk,
                input=True)

frames = []  # Initialize array to store frames

# Store data in chunks for 3 seconds
for i in range(0, int(fs / chunk * seconds)):
    data = stream.read(chunk)
    frames.append(data)

# Stop and close the stream 
stream.stop_stream()
stream.close()
# Terminate the PortAudio interface
p.terminate()

print('Finished recording')

# Save the recorded data as a WAV file
wf = wave.open(filename, 'wb')
wf.setnchannels(channels)
wf.setsampwidth(p.get_sample_size(sample_format))
wf.setframerate(fs)
wf.writeframes(b''.join(frames))
wf.close()
```
  ,it throws this error
`Traceback (most recent call last):
  File "/data/data/com.termux/files/home/main1.py", line 15, in <module>
    stream = p.open(format=sample_format,
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/pyaudio/__init__.py", line 639, in open
    stream = PyAudio.Stream(self, *args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/pyaudio/__init__.py", line 441, in __init__
    self._stream = pa.open(**arguments)
                   ^^^^^^^^^^^^^^^^^^^^
OSError: [Errno -9996] Invalid input device (no default output device)`

### What is the expected behavior?

It should work without rasing any error 

### System information

```shell
Termux Variables:
TERMUX_API_VERSION=0.50.1
TERMUX_APK_RELEASE=F_DROID
TERMUX_APP_PACKAGE_MANAGER=apt
TERMUX_APP_PID=31431
TERMUX_IS_DEBUGGABLE_BUILD=0
TERMUX_MAIN_PACKAGE_FORMAT=debian
TERMUX_VERSION=0.118.0
TERMUX__USER_ID=0
Packages CPU architecture:
aarch64
Subscribed repositories:
# sources.list
deb https://packages-cf.termux.dev/apt/termux-main stable main
# x11-repo (sources.list.d/x11.list)
deb https://packages-cf.termux.dev/apt/termux-x11 x11 main
# tur-repo (sources.list.d/tur.list)
deb https://tur.kcubeterm.com tur-packages tur tur-on-device tur-continuous
# root-repo (sources.list.d/root.list)
deb https://packages-cf.termux.dev/apt/termux-root root stable
# sources.list.d/termux-adb.list
deb https://nohajc.github.io termux extras
Updatable packages:
neovim/stable 0.10.1-1 aarch64 [upgradable from: 0.10.1]
openssh-sftp-server/stable 9.9p1-3 aarch64 [upgradable from: 9.9p1-2]
openssh/stable 9.9p1-3 aarch64 [upgradable from: 9.9p1-2]
termux-tools version:
1.43.5
Android version:
10
Kernel build information:
Linux localhost 4.9.186-perf-g131f907 #1 SMP PREEMPT Thu Apr 8 17:20:44 WIB 2021 aarch64 Android
Device manufacturer:
Xiaomi
Device model:
Redmi Y3
LD Variables:
LD_LIBRARY_PATH=
LD_PRELOAD=/data/data/com.termux/files/usr/lib/libtermux-exec.so
Installed termux plugins:
com.termux.api versionCode:51
```

### PR DESCRIPTION
**Pull request by twaik**
Fixes #18790
Fixes #10412
Fixes #21545
Fixes #20570
Fixes #19679
Fixes #5422

Audacity starts, can open wav and mp3 files, play and record sound (with RECORD_AUDIO permission granted to Termux:API). Some review needed.

### CODE REVIEW

**Comment by Biswa96** [comment]
I have tested the aarch64 CI artifact in my phone but got an error.

```
CANNOT LINK EXECUTABLE "audacity": library "lib-theme-resources.so" not found: needed by main executable
```

After exporting the variables like this file https://github.com/audacity/audacity/blob/master/au3/linux/audacity.sh

```
$ LD_LIBRARY_PATH=$PREFIX/lib/audacity AUDACITY_MODULES_PATH=$PREFIX/lib/audacity/modules audacity
Bad system call
```
 audacity fails with semget syscall.

**Comment by twaik** [comment]
Ok, so I removed usage of semget/semctl/semop and also shmget and managed to make it work without them. Also I added `-Wl,-rpath` to make it work without LD_LIBRARY_PATH. Should work now.

**Comment by twaik** [comment]
By `should work` I meant the patch worked fine with audacity built for regular linux, I did not check it on my own device.

**Comment by Biswa96** [comment]
Now audacity opens does not play any audio file. Try to import a mp3 file and click on the play button.

**Comment by twaik** [comment]
I tried to open m4a file and it refused to open it too. Probably it can not pick ffmpeg.

**Comment by twaik** [comment]
Interesting. The same happens to old audacity, probably portaudio related.

**Comment by twaik** [comment]
[portaudio.tar.gz](https://github.com/user-attachments/files/17401084/portaudio.tar.gz)
Fun. I managed to make portaudio with opensles using some old sources, so output works, but no input.
And this portaudio crashes after a finishing playing my sample sound, probably some bug in portaudio or in opensles hostapi backend.

**Comment by twaik** [comment]
Sample to enumerate portaudio devices.
```
#include <stdio.h>
#include "portaudio.h"

int main(void)
{
    PaError err;

    /* Initialize PortAudio */
    err = Pa_Initialize();
    if (err != paNoError) {
        fprintf(stderr, "PortAudio error during initialization: %s\n", Pa_GetErrorText(err));
        return 1;
    }

    /* Get the number of audio devices */
    int numDevices = Pa_GetDeviceCount();
    if (numDevices < 0) {
        fprintf(stderr, "PortAudio error getting device count: %s\n", Pa_GetErrorText(numDevices));
        Pa_Terminate();
        return 1;
    }

    printf("Number of devices found: %d\n\n", numDevices);

    /* Iterate over each device */
    for (int i = 0; i < numDevices; i++) {
        const PaDeviceInfo *deviceInfo = Pa_GetDeviceInfo(i);

        printf("Device #%d:\n", i);
        printf("  Name                        : %s\n", deviceInfo->name);
        printf("  Host API                    : %s\n", Pa_GetHostApiInfo(deviceInfo->hostApi)->name);
        printf("  Max input channels          : %d\n", deviceInfo->maxInputChannels);
        printf("  Max output channels         : %d\n", deviceInfo->maxOutputChannels);
        printf("  Default sample rate         : %.2f\n", deviceInfo->defaultSampleRate);
        printf("  Default low input latency   : %.4f\n", deviceInfo->defaultLowInputLatency);
        printf("  Default low output latency  : %.4f\n", deviceInfo->defaultLowOutputLatency);
        printf("  Default high input latency  : %.4f\n", deviceInfo->defaultHighInputLatency);
        printf("  Default high output latency : %.4f\n", deviceInfo->defaultHighOutputLatency);
        printf("\n");
    }

    /* Terminate PortAudio */
    err = Pa_Terminate();
    if (err != paNoError) {
        fprintf(stderr, "PortAudio error during termination: %s\n", Pa_GetErrorText(err));
        return 1;
    }

    return 0;
}
```

**Comment by twaik** [comment]
Ok, I figured out. For some reason `pthread_join` crashed application instead returning immediately or throwing `ESRCH` error. Now sound output works.

**Comment by twaik** [comment]
AFAIK microphone support is not possible without RECORD_AUDIO so it will be `wontfix` until @agnostic-apollo decides to add it to termux-app's application manifest.

**Comment by agnostic-apollo** [comment]
You can install termux-api, will have to see about adding to termux-app itself.

https://github.com/termux/termux-api/blob/c553bc2ba7de37ad5c42df23d71e699d2785d7c5/app/src/main/AndroidManifest.xml#L27

**Comment by twaik** [comment]
Wow! So microphone works fine too! So cool!
I did not know termux:api can give permissions to tools of main termux app.

**Comment by agnostic-apollo** [comment]
They share  `sharedUserId`, so permissions are shared between all apps under the same uid regardless of who requests it.

**Comment by twaik** [comment]
I'll merge this PR in 24 hours if no one minds.

**Comment by agnostic-apollo** [comment]
Oh, so you finally have the audacity to merge! huh!

**Comment by truboxl** [comment]
Should turn off networking, update checks and crash reports

File: x11-packages/audacity/build.sh [hunk 1]
```diff
@@ -2,47 +2,116 @@ TERMUX_PKG_HOMEPAGE=https://www.audacityteam.org/
...
 TERMUX_PKG_SUGGESTS="audacity-ffmpeg"
+TERMUX_PKG_HOSTBUILD=true
 TERMUX_PKG_EXTRA_CONFIGURE_ARGS="
+-DCMAKE_STRIP=llvm-strip
+-Daudacity_conan_enabled=off
+-Daudacity_has_vst3=no
+-Daudacity_has_tests=no
+-Daudacity_has_networking=no
+-Daudacity_has_crashreports=no
+-Daudacity_has_sentry_reporting=no
```
Comment (truboxl) [inline x11-packages/audacity/build.sh:23]: Missing `-Daudacity_has_updates_check=off`

Comment (twaik) [reply x11-packages/audacity/build.sh:23]: Done.

**Comment by twaik** [comment]
Fixed a sqlite-related problem. 
```
 $ audacity
 CANNOT LINK EXECUTABLE "audacity": cannot locate symbol "sqlite3_trace" referenced by "/system/lib64/libandroid_runtime.so"...
```

**Pull request merged**